include/aos/iam/nodemanager/nodemanager.hpp

```diff
@@ -0,0 +1,251 @@
+/*
+ * Copyright (C) 2024 Renesas Electronics Corporation.
+ * Copyright (C) 2024 EPAM Systems, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef AOS_NODEMANAGER_HPP_
+#define AOS_NODEMANAGER_HPP_
+
+#include "aos/common/types.hpp"
+#include "aos/iam/config.hpp"
+
+namespace aos::iam::nodemanager {
+
+/** @addtogroup IAM Node Manager
+ *  @{
+ */
+
+/**
+ * NodeInfo listener interface.
+ */
+class NodeInfoListenerItf {
+public:
+    /**
+     * Node info change notification.
+     *
+     * @param info node info.
+     */
+    virtual void OnNodeInfoChange(const NodeInfo& info) = 0;
+
+    /**
+     * Node info removed notification.
+     *
+     * @param id id of the node been removed.
+     */
+    virtual void OnNodeRemoved(const String& id) = 0;
+
+    /**
+     * Destroys object instance.
+     */
+    virtual ~NodeInfoListenerItf() = default;
+};
+
+/**
+ * Node manager interface.
+ */
+class NodeManagerItf {
+public:
+    /**
+     * Updates whole information for a node.
+     *
+     * @param info node info.
+     * @return Error.
+     */
+    virtual Error SetNodeInfo(const NodeInfo& info) = 0;
+
+    /**
+     * Updates node status.
+     *
+     * @param nodeId node identifier.
+     * @param status node status.
+     * @return Error.
+     */
+    virtual Error SetNodeStatus(const String& nodeId, NodeStatus status) = 0;
+
+    /**
+     * Returns node info.
+     *
+     * @param nodeId node identifier.
+     * @param[out] nodeInfo result node identifier.
+     * @return Error.
+     */
+    virtual Error GetNodeInfo(const String& nodeId, NodeInfo& nodeInfo) const = 0;
+
+    /**
+     * Returns ids for all the node in the manager.
+     *
+     * @param ids result node identifiers.
+     * @return Error.
+     */
+    virtual Error GetAllNodeIds(Array<StaticString<cNodeIDLen>>& ids) const = 0;
+
+    /**
+     * Removes node info by its id.
+     *
+     * @param nodeId node identifier.
+     * @return Error.
+     */
+    virtual Error RemoveNodeInfo(const String& nodeId) = 0;
+
+    /**
+     * Subscribes listener for node info updates.
+     *
+     * @param listener listener to subscribe.
+     * @return Error.
+     */
+    virtual Error SubscribeNodeInfoChange(NodeInfoListenerItf& listener) = 0;
+
+    /**
+     * Destroys object instance.
+     */
+    virtual ~NodeManagerItf() = default;
+};
+
+/**
+ * Node info storage interface.
+ */
+class NodeInfoStorageItf {
```

Comment:
Better to use CRUD operations for storage. We do not always have SQL as storage and set individual fields in storage may require more effort than just update it on client side. So more simple and usable API will look like:

```c++
AddNodeInfo
UpdateNodeInfo
GetNodeInfo
GetAllNodeIds
RemoveNodeInfo
```

Reply:
making a crud interface makes it usage more difficult for client. I would better hide complexity inside an implementation behind a simpler interface.

Reply:
agreed to remove status(get/set) methods